### PR TITLE
[Workspace]Optional workspaces params in repository

### PR DESCRIPTION
**Pull request by SuZhou-Joe**
### Description

As workspace will be a new concept to organize saved objects, this PR add a new field `workspaces` to indicate which workspaces the objects belong to and consume the field in related methods.

- [x] Add workspaces field in saved objects mappings
- [x] Consume worksapces when create object
- [x] Consume workspaces when bulkCreate objects
- [x] Consume workspaces when checkConflicts
- [x] Consume workspaces when validating reference

### Issues Resolved

Partial #4944 

## Screenshot

N/A

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by SuZhou-Joe** [comment]
#4944

**Comment by kavilla** [comment]
Just a heads up @SuZhou-Joe, modified the PR description.

if it reads `closes #issue number` it will auto attach the issue and then the issue will be closed on merged. even if there is a word before `closes`.

Not a big deal, can always re-open the issue if it does close.

**Comment by SuZhou-Joe** [comment]
> Just a heads up @SuZhou-Joe, modified the PR description.
> 
> if it reads `closes #issue number` it will auto attach the issue and then the issue will be closed on merged. even if there is a word before `closes`.
> 
> Not a big deal, can always re-open the issue if it does close.

Thanks @kavilla , modified the PR description.

**Review by zengyan-amazon — COMMENTED** [review]
for all these workspace operations, can we do it with interceptors like `savedObjectClientWrapper` instead of making surgical changes the repository deeply?

File: src/core/server/saved_objects/export/get_sorted_objects_for_export.ts [hunk 1]
```diff
@@ -121,6 +127,7 @@ async function fetchObjectsToExport({
       search,
       perPage: exportSizeLimit,
       namespaces: namespace ? [namespace] : undefined,
+      ...(workspaces ? { workspaces } : {}),
```
[the remark this answers is not in the document]
Comment (SuZhou-Joe) [reply src/core/server/saved_objects/export/get_sorted_objects_for_export.ts:130]: Workspaces is an array, in here if workspaces is null, `...({})` will have no impact on the params that passed to find, if we use `workspaces: workspaces ? workspaces : undefined`, there will be a etra `{ ...., workspaces: undefined }` in find params.

**Comment by SuZhou-Joe** [comment]
> for all these workspace operations, can we do it with interceptors like `savedObjectClientWrapper` instead of making surgical changes the repository deeply?

You mean the preflightCheck on conflict with workspace right? That's a good advice and I haven't think about it. The changes made to repository.ts can be divided into two parts:
1. Consume the workspaces param and pass it to serializer / make serializer consume `workspaces` from input params or from .kibana index.
2. Some preflight check on workspaces field and target workspace, so that the workspaces field won't be overwrite by accident. 

For part 2, I think it is doable and has much benefit in the long term, will try to make the refractor.

**Comment by SuZhou-Joe** [comment]
@zengyan-amazon I did the refactor and you can find the changes in https://github.com/SuZhou-Joe/OpenSearch-Dashboards/pull/8 , please take a look to see if it is what you want, if so I will merge the refactor to this branch.

**Comment by SuZhou-Joe** [comment]
Closed because of #5949.

**Pull request closed without merging**